### PR TITLE
Add client template based on Vue.js

### PR DESCRIPTION
**Pull request by jortizsao**
It's based on the vue-cli webpack template (https://github.com/vuejs-templates/webpack) and includes among other things:

* Jest
* Babel
* Webpack
* ESLint

### CODE REVIEW

File: client/package.json [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    "vue-loader": "^13.3.0",
+    "vue-style-loader": "^3.0.1",
+    "vue-template-compiler": "^2.5.2",
+    "webpack": "^3.6.0",
+    "webpack-bundle-analyzer": "^2.9.0",
+    "webpack-dev-server": "^2.9.1",
+    "webpack-merge": "^4.1.0"
+  },
+  "engines": {
+    "node": ">= 6.0.0",
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/package.json]: we can use node >=8 (current LTS) if you want. I change it

File: client/build/utils.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        fallback: 'vue-style-loader'
+      })
+    } else {
+      return ['vue-style-loader'].concat(loaders)
+    }
+  }
+
+  // https://vue-loader.vuejs.org/en/configurations/extract-css.html
+  return {
+    css: generateLoaders(),
```
Comment (lojzatran) [inline client/build/utils.js]: What's the difference between css and postcss?

File: client/build/utils.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    postcss: generateLoaders(),
+    less: generateLoaders('less'),
+    sass: generateLoaders('sass', { indentedSyntax: true }),
+    scss: generateLoaders('sass'),
```
Comment (lojzatran) [inline client/build/utils.js]: Should here be scss?

File: client/build/utils.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      return ExtractTextPlugin.extract({
+        use: loaders,
+        fallback: 'vue-style-loader'
+      })
+    } else {
+      return ['vue-style-loader'].concat(loaders)
+    }
+  }
+
+  // https://vue-loader.vuejs.org/en/configurations/extract-css.html
```
Comment (lojzatran) [inline client/build/utils.js]: How is this link related to the return statement below?

File: client/build/utils.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+      test: new RegExp('\\.' + extension + '$'),
+      use: loader
+    })
+  }
+
+  return output
+}
+
+exports.createNotifierCallback = () => {
+  const notifier = require('node-notifier')
```
Comment (lojzatran) [inline client/build/utils.js]: Why not put `require` to the beginning of the file? It's then easier to see what other modules this file needs.

File: client/.eslintrc.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
+// https://eslint.org/docs/user-guide/configuring
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/.eslintrc.js]: ok, I change it

File: client/build/check-versions.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+const chalk = require('chalk')
+const semver = require('semver')
+const packageConfig = require('../package.json')
+const shell = require('shelljs')
+
+function exec (cmd) {
+  return require('child_process').execSync(cmd).toString().trim()
+}
+
+const versionRequirements = [
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/check-versions.js]: It's coming from the official Vue.js webpack template (https://github.com/vuejs-templates/webpack/blob/develop/template/build/check-versions.js). 

it checks the version of your node/npm, and compare them to the ones set in the engines section within the package.json. If it's not met, logs out a warning

[the remark this answers is not in the document]
Comment (lojzatran) [reply client/build/check-versions.js]: Can you write this in the file?

File: client/build/utils.js (hunk 3 above)
Comment (jortizsao) [reply client/build/utils.js]: 'sass' and 'scss' are both syntaxes of Sass. 

sass is the older syntax and scss is the new syntax, and that's why both syntaxes use the same loaders

File: client/build/check-versions.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/check-versions.js]: sure!

File: client/config/dev.env.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+'use strict'
+const merge = require('webpack-merge')
+const prodEnv = require('./prod.env')
+
+module.exports = merge(prodEnv, {
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/config/dev.env.js]: I agree with you. It's also coming from the Vue.js official template. I have been thinking the reason why they do this way, and I haven't found a good reason (I have also looked at the PR's in the original  template file in order to find a comment regarding this)

Maybe @butenkor or @heshamMassoud know a good reason.

File: client/package.json [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+    "eslint-friendly-formatter": "^3.0.0",
+    "eslint-import-resolver-webpack": "^0.8.3",
+    "eslint-loader": "^1.7.1",
+    "eslint-plugin-import": "^2.7.0",
+    "eslint-plugin-vue": "^4.0.0",
+    "extract-text-webpack-plugin": "^3.0.0",
+    "file-loader": "^1.1.4",
+    "friendly-errors-webpack-plugin": "^1.6.1",
+    "html-webpack-plugin": "^2.30.1",
+    "jest": "^22.0.4",
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/package.json]: do you know why they were not satisfied?

I'm using it for one year or so and I'm very happy with it. Jest has become the "standard" libraries for the frontend frameworks (React, Vue.js, etc...). 

It includes features that are really helpful for testing components like Snapshot testing (https://facebook.github.io/jest/docs/en/snapshot-testing.html). 

And you don't have to use Istanbul for coverage (coverage is included in Jest) nor an mock/spy library like Sinon because they are also included.

I know that in commercetools is also used by the MC team (here you can find an article from Nicola Molinari https://techblog.commercetools.com/testing-in-react-best-practices-tips-and-tricks-577bb98845cd)

I got used to Jasmine and the shift to Jest was really smooth. The syntax is easy and familiar if you have used other testing frameworks. And if you have any trouble or issue, I will be more than happy to help you :)

[the remark this answers is not in the document]
Comment (jortizsao) [reply client/package.json]: mmm... I haven't suffered the issue you describe. 

In order to keep the consistency within the project, I also set Jest as the framework in the backend, so the same developer with little effort can work in the frontend and the backend.

File: client/.eslintrc.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/.eslintrc.js]: @lojzatran I have added the eslint commercetools ruleset. One question that I have is that I saw the .eslintrc of the project https://github.com/commercetools/commercetools-node-variant-reassignment/blob/master/.eslintrc

and I have seen that you have overwritten the behavior of some of the rules. What's the criteria to do this? it's a common eslint config for all commercetools project? if so, why is not included in the commercetools ruleset?

File: client/build/build.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+const path = require('path')
+const chalk = require('chalk')
+const webpack = require('webpack')
+const config = require('../config')
+const webpackConfig = require('./webpack.prod.conf')
+
+const spinner = ora('building for production...')
+spinner.start()
+
+rm(path.join(config.build.assetsRoot, config.build.assetsSubDirectory), err => {
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/build.js]: ok, I write it down to clarify.

Basically, webpack creates the different bundles (.js, . css, .map, etc...) and if error then it throws it out, and if success it shows you the statistics (the bundles that have been created, the size of the each bundle, and the order of the chunks)

File: client/build/utils.js [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+
+exports.assetsPath = function (_path) {
+  const assetsSubDirectory = process.env.NODE_ENV === 'production'
+    ? config.build.assetsSubDirectory
+    : config.dev.assetsSubDirectory
+
+  return path.posix.join(assetsSubDirectory, _path)
+}
+
+exports.cssLoaders = function (options) {
```
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/utils.js]: I'll do it

File: client/build/utils.js (hunk 4 above)
Comment (jortizsao) [reply client/build/utils.js]: it's not related to the return statement below. It's related to the webpack ExtractTextPlugin above.

I move the link up close to the ExtractTextPlugin in order to clarify.

File: client/build/utils.js (hunk 2 above)
Comment (jortizsao) [reply client/build/utils.js]: PostCSS allows to use Javascript for processing your CSS

Honestly, I've never used but I have seen people using it via snippets or functions, when they need to deal with the CSS in a more 'programmatic' way

File: client/build/utils.js (hunk 5 above)
Comment (jortizsao) [reply client/build/utils.js]: This file is also based on the official vue.js template.

It's because depending on your config you will use or not this createNotifierCallback. In this particular case, it's used only in "dev" mode (it's only called by the webpack.dev.conf.js file) but not in "prod" mode.

Since you don't want to include unnecessary libraries in your bundle (in the backend you don't care but in the frontend the bundle size is important for speeding up your loading time), the import of the library is just made once the function is used. If you put the library at the beginning, it will be imported for "dev" and "prod" modes

File: client/.eslintrc.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (lojzatran) [reply client/.eslintrc.js]: Yes, we didn't like some of the rules, so we add exceptions :D You will see by yourself that some rules are not good, so you can change it. But you don't have to specify all the rules as you did before. For example, we want to make the line longer, so we override this rule: `"max-len": ["error", 100]`.

 The eslint file above is not common for all projects.

[the remark this answers is not in the document]
Comment (jortizsao) [reply client/.eslintrc.js]: so what should I do? keep the commercetools ruleset as it is? overwrite some rules? :)

I mean, if the goal of using the commercetools ruleset rather than the airbnb ruleset, is to encourage all commercetools developers to program the same way, but then if each team overwrites these rules as their like. 

what should we do in a new project like this?

File: client/build/build.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/build.js]: https://github.com/commercetools/kubernetes-custom-ui/pull/1/commits/fecda26bec8660e1145383678daf00495f508768

File: client/build/check-versions.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/check-versions.js]: https://github.com/commercetools/kubernetes-custom-ui/pull/1/commits/fecda26bec8660e1145383678daf00495f508768

File: client/build/utils.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (jortizsao) [reply client/build/utils.js]: https://github.com/commercetools/kubernetes-custom-ui/pull/1/commits/fecda26bec8660e1145383678daf00495f508768

File: client/build/utils.js (hunk 4 above)
Comment (jortizsao) [reply client/build/utils.js]: https://github.com/commercetools/kubernetes-custom-ui/pull/1/commits/fecda26bec8660e1145383678daf00495f508768

File: client/.eslintrc.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (junajan) [reply client/.eslintrc.js]: I would say that it is a good thing if we stick to one code style but CTs eslint config can be outdated so we should not take it too strictly

[the remark this answers is not in the document]
Comment (lojzatran) [reply client/.eslintrc.js]: You can try to stick to the CT rules. With time, you will see if you need to add exceptions or not, and then we can discuss.
